### ISSUE TITLE
Native buffer optimization in C#

### ISSUE THREAD

**Issue by Gusi** [sjoerdvankreel/xt-audio#7]
In the example code dealing with native buffers, I think the intermediate copy to a byte array can be avoided accessing the raw buffer through a Span struct. As the Span in this case is created in the stack, I don't foresee any performance penalty.

Original code

```c#
     static int OnInterleavedNativeBuffer(XtStream stream, in XtBuffer buffer, object user)
    {
        var ctx = (Context)user;
        int bytes = GetBufferSize(Channels.inputs, buffer.frames);
        Marshal.Copy(buffer.input, ctx.intermediate, 0, bytes);
        ctx.recording.Write(ctx.intermediate, 0, bytes);
        return 0;
    }
```

Proposed code

```c#
    static unsafe int OnInterleavedNativeBuffer(XtStream stream, in XtBuffer buffer, object user)
    {
        var ctx = (Context)user;
        int bytes = GetBufferSize(Channels.inputs, buffer.frames);
        var aux = new Span<byte>((byte*)buffer.input, bytes);
        ctx.recording.Write(aux);
        return 0;
    }
```

**Comment by sjoerdvankreel** [comment]
If i'm not mistaken, span is not available for .NET framework? Otherwise it might be better to drop the managed/native distinction entirely and go all the way with Span. Make input a ReadOnlySpan<byte> and output a Span<byte> on XtBuffer directly. Should be possibly as XtBuffer is a struct already. Would have to check how to expose non-interleaved buffers and do something similar for java, probably using DirectByteBuffer and friends. But again, i think this is only possible on .net core/5.

**Comment by Gusi** [comment]
You are completely right about Span only being available in Core 2.1+ and .Net 5+ versions. I suppose that many developers are moving to .Net 5 and deprecating .Net Framework support, so maybe it could be useful at least to let them know about this little optimization even if it is not 100% usable in the whole .NET ecosystem. In my case, I just discovered some of these new tricks when moving to .Net 5, like Span, stackalloc and others...

**Comment by sjoerdvankreel** [comment]
I'm gonna postpone this as I want to keep parity with NET framework for a bit longer. Performance-minded users can always use Span directly on top of native buffers (as you correctly demonstrated) or use IntPtr directly so there's no real loss here. Closing this for now, but thanks for the tip though.
